pkg/apis/stackconfigpolicy/v1alpha1/stackconfigpolicy_types.go

```diff
@@ -0,0 +1,210 @@
+	Phase PolicyPhase `json:"phase,omitempty"`
+	// ObservedGeneration is the most recent generation observed for this StackConfigPolicy.
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+}
+
+type PolicyPhase string
+
+const (
+	UnknownPhase         PolicyPhase = "Unknown"
+	ReadyPhase           PolicyPhase = "Ready"
+	ApplyingChangesPhase PolicyPhase = "ApplyingChanges"
+	InvalidPhase         PolicyPhase = "Invalid"
+	ErrorPhase           PolicyPhase = "Error"
+	ConflictPhase        PolicyPhase = "Conflict"
+)
+
+// phaseOrder
+var phaseOrder = map[PolicyPhase]int{
+	UnknownPhase:         -1,
+	ReadyPhase:           0,
+	ApplyingChangesPhase: 1,
+	InvalidPhase:         2,
+	ErrorPhase:           3,
+	ConflictPhase:        4,
+}
+
+// ResourcePolicyStatus models the status of the policy for one resource to be configured.
+type ResourcePolicyStatus struct {
+	Phase           PolicyPhase       `json:"phase,omitempty"`
+	CurrentVersion  int64             `json:"currentVersion,omitempty"`
+	ExpectedVersion int64             `json:"expectedVersion,omitempty"`
+	Error           PolicyStatusError `json:"error,omitempty"`
+}
+
+type PolicyStatusError struct {
+	Version int64    `json:"version,omitempty"`
+	Errors  []string `json:"errors,omitempty"`
+}
+
+func NewStatus(scp StackConfigPolicy) StackConfigPolicyStatus {
+	status := StackConfigPolicyStatus{
+		ResourcesStatuses:  map[string]ResourcePolicyStatus{},
+		Phase:              ReadyPhase,
+		ObservedGeneration: scp.Generation,
+	}
+	status.setReadyCount()
+	return status
+}
+
+func (s *StackConfigPolicyStatus) setReadyCount() {
+	s.ReadyCount = fmt.Sprintf("%d/%d", s.Ready, s.Resources)
+}
+
+func (s *StackConfigPolicyStatus) AddPolicyErrorFor(resource types.NamespacedName, phase PolicyPhase, msg string) {
+	s.ResourcesStatuses[resource.String()] = ResourcePolicyStatus{
+		Phase: phase,
+		Error: PolicyStatusError{Errors: []string{msg}},
+	}
+	s.Update()
+}
+
+func (s *StackConfigPolicyStatus) UpdateResourceStatusPhase(resource types.NamespacedName, status ResourcePolicyStatus) {
+	if status.CurrentVersion == unknownVersion { //nolint:gocritic
+		status.Phase = UnknownPhase
+	} else if status.Error.Errors != nil {
+		status.Phase = ErrorPhase
+	} else if status.CurrentVersion == status.ExpectedVersion {
+		status.Phase = ReadyPhase
+	} else {
+		status.Phase = ApplyingChangesPhase
+	}
+	s.ResourcesStatuses[resource.String()] = status
+	s.Update()
+}
+
+// Update updates the policy status from its resources statuses.
+func (s *StackConfigPolicyStatus) Update() {
+	s.Resources = len(s.ResourcesStatuses)
+	s.Ready = 0
+	s.Error = 0
+	for _, status := range s.ResourcesStatuses {
+		resourcePhase := status.Phase
+		if resourcePhase == ReadyPhase {
+			s.Ready++
+		} else if resourcePhase == ErrorPhase {
+			s.Error++
+		}
+		// update phase if that of the resource status is worse
+		if phaseOrder[resourcePhase] > phaseOrder[s.Phase] {
+			s.Phase = resourcePhase
+		}
```

Comment:
So the StackConfigPolicy.Phase is always matching the worst phase of all of its managed resources?  Is that what is intended here?

Reply:
Exactly.